File: routes/personnel_routes.py
"""
Routes cho quản lý quân số và cắt cơm — Multi-tenant
"""
from flask import Blueprint, jsonify, request
from flask_login import login_required
from database import db, ChienSi, DonVi, CatCom, get_user_don_vi_id
from datetime import date, datetime

personnel_bp = Blueprint('personnel', __name__)


@personnel_bp.route('/api/personnel')
@login_required
def get_personnel():
    """Lấy danh sách quân nhân (theo đơn vị)"""
    dvid = get_user_don_vi_id()
    don_vi_id = request.args.get('don_vi_id', type=int)
    
    # Lấy đơn vị con
    don_vi_ids = []
    if dvid:
        don_vi_ids = [dvid]
        children = DonVi.query.filter_by(don_vi_cha_id=dvid).all()
        don_vi_ids.extend([c.id for c in children])
    
    query = ChienSi.query.filter(ChienSi.don_vi_id.in_(don_vi_ids))
    if don_vi_id and don_vi_id in don_vi_ids:
        query = ChienSi.query.filter_by(don_vi_id=don_vi_id)
    return jsonify([cs.to_dict() for cs in query.all()])


@personnel_bp.route('/api/personnel/stats')
@login_required
def get_personnel_stats():
    """Thống kê quân số (theo đơn vị)"""
    dvid = get_user_don_vi_id()
    today = date.today()
    
    don_vi_ids = []
    if dvid:
        don_vi_ids = [dvid]
        children = DonVi.query.filter_by(don_vi_cha_id=dvid).all()
        don_vi_ids.extend([c.id for c in children])
    
    base_query = ChienSi.query.filter(ChienSi.don_vi_id.in_(don_vi_ids))
    total = base_query.count()
    tai_vi = base_query.filter(ChienSi.trang_thai == 'tai_vi').count()
    vang_mat = total - tai_vi
    
    # Số cắt cơm active hôm nay (chiến sĩ trong đơn vị)
    cat_com_today = CatCom.query.join(ChienSi).filter(
        CatCom.ngay_bat_dau <= today,
        CatCom.ngay_ket_thuc >= today,
        ChienSi.don_vi_id.in_(don_vi_ids)
    ).count()
    
    return jsonify({
        'tong_quan_so': total,
        'tai_don_vi': tai_vi,
        'vang_mat': vang_mat,
        'cat_com_hom_nay': cat_com_today,
        'an_tai_don_vi': total - cat_com_today
    })


@personnel_bp.route('/api/units')
@login_required
def get_units():
    """Lấy danh sách đơn vị con"""
    dvid = get_user_don_vi_id()
    if dvid:
        units = DonVi.query.filter(
            db.or_(DonVi.id == dvid, DonVi.don_vi_cha_id == dvid)
        ).all()
    else:
        units = []
    return jsonify([u.to_dict() for u in units])


@personnel_bp.route('/api/meal-cuts')
@login_required
def get_meal_cuts():
    """Lấy danh sách cắt cơm (theo đơn vị)"""
    dvid = get_user_don_vi_id()
    today = date.today()
    active_only = request.args.get('active', 'true').lower() == 'true'
    
    don_vi_ids = [dvid] if dvid else []
    if dvid:
        children = DonVi.query.filter_by(don_vi_cha_id=dvid).all()
        don_vi_ids.extend([c.id for c in children])
    
    query = CatCom.query.join(ChienSi).filter(ChienSi.don_vi_id.in_(don_vi_ids))
    if active_only:
        query = query.filter(CatCom.ngay_bat_dau <= today, CatCom.ngay_ket_thuc >= today)
    return jsonify([cc.to_dict() for cc in query.order_by(CatCom.ngay_tao.desc()).all()])


@personnel_bp.route('/api/meal-cuts', methods=['POST'])
@login_required
def create_meal_cut():
    """Tạo báo cắt cơm"""
    data = request.json
    cc = CatCom(
        chien_si_id=data['chien_si_id'],
        ngay_bat_dau=datetime.strptime(data['ngay_bat_dau'], '%Y-%m-%d').date(),
        ngay_ket_thuc=datetime.strptime(data['ngay_ket_thuc'], '%Y-%m-%d').date(),
        ly_do=data.get('ly_do', ''),
        loai=data.get('loai', 'cat_com'),
        ghi_chu=data.get('ghi_chu'),
        nguoi_bao=data.get('nguoi_bao')
    )
    cs = ChienSi.query.get(data['chien_si_id'])
    if cs:
        if cc.ly_do == 'phep': cs.trang_thai = 'phep'
        elif cc.ly_do == 'gac': cs.trang_thai = 'gac'
        else: cs.trang_thai = 'cong_tac'
    
    db.session.add(cc)
    db.session.commit()
    return jsonify(cc.to_dict()), 201


@personnel_bp.route('/api/meal-cuts/<int:cut_id>', methods=['DELETE'])
@login_required
def delete_meal_cut(cut_id):
    """Hủy cắt cơm"""
    cc = CatCom.query.get_or_404(cut_id)
    cs = ChienSi.query.get(cc.chien_si_id)
    if cs:
        cs.trang_thai = 'tai_vi'
    db.session.delete(cc)
    db.session.commit()
    return jsonify({'message': 'Đã hủy cắt cơm'})


@personnel_bp.route('/api/personnel/birthdays')
@login_required
def get_birthdays():
    """Lấy danh sách chiến sĩ có sinh nhật trong tháng (theo đơn vị)"""
    dvid = get_user_don_vi_id()
    month = request.args.get('month', type=int, default=date.today().month)
    from sqlalchemy import extract
    
    don_vi_ids = [dvid] if dvid else []
    if dvid:
        children = DonVi.query.filter_by(don_vi_cha_id=dvid).all()
        don_vi_ids.extend([c.id for c in children])
    
    cs_list = ChienSi.query.filter(
        extract('month', ChienSi.ngay_sinh) == month,
        ChienSi.don_vi_id.in_(don_vi_ids)
    ).all()
    return jsonify([cs.to_dict() for cs in cs_list])

File: routes/personnelRoutes.js
/**
 * Personnel routes - Quân số & Cắt cơm
 */
const express = require('express');
const router = express.Router();
const { db, getUserDonViId, getDonViChildIds, chienSiToDict, catComToDict, donViToDict } = require('../database');

const loginRequired = (req, res, next) => {
  if (!req.session?.user) return res.status(401).json({ error: 'Chua dang nhap' });
  next();
};

router.get('/api/personnel', loginRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const dvIdParam = req.query.don_vi_id ? parseInt(req.query.don_vi_id) : null;
    const donViIds = await getDonViChildIds(dvid);
    if (!donViIds.length) return res.json([]);
    if (dvIdParam && donViIds.includes(dvIdParam)) {
      const list = await db.all('SELECT * FROM chien_si WHERE don_vi_id = ?', dvIdParam);
      return res.json(await Promise.all(list.map(chienSiToDict)));
    }
    const ph = donViIds.map(() => '?').join(',');
    const list = await db.all(`SELECT * FROM chien_si WHERE don_vi_id IN (${ph})`, ...donViIds);
    res.json(await Promise.all(list.map(chienSiToDict)));
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.get('/api/personnel/stats', loginRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const today = new Date().toISOString().split('T')[0];
    const donViIds = await getDonViChildIds(dvid);
    if (!donViIds.length) return res.json({ tong_quan_so: 0, tai_don_vi: 0, vang_mat: 0, cat_com_hom_nay: 0, an_tai_don_vi: 0 });
    const ph = donViIds.map(() => '?').join(',');
    const total = (await db.get(`SELECT COUNT(*) as c FROM chien_si WHERE don_vi_id IN (${ph})`, ...donViIds)).c;
    const taiVi = (await db.get(`SELECT COUNT(*) as c FROM chien_si WHERE don_vi_id IN (${ph}) AND trang_thai = 'tai_vi'`, ...donViIds)).c;
    const catComToday = (await db.get(`SELECT COUNT(*) as c FROM cat_com cc JOIN chien_si cs ON cc.chien_si_id = cs.id
      WHERE cc.ngay_bat_dau <= ? AND cc.ngay_ket_thuc >= ? AND cs.don_vi_id IN (${ph})`, today, today, ...donViIds)).c;
    const t = parseInt(total) || 0;
    const tv = parseInt(taiVi) || 0;
    const cc = parseInt(catComToday) || 0;
    res.json({ tong_quan_so: t, tai_don_vi: tv, vang_mat: t - tv, cat_com_hom_nay: cc, an_tai_don_vi: t - cc });
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.get('/api/units', loginRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    if (!dvid) return res.json([]);
    const units = await db.all('SELECT * FROM don_vi WHERE id = ? OR don_vi_cha_id = ?', dvid, dvid);
    res.json(await Promise.all(units.map(donViToDict)));
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.get('/api/meal-cuts', loginRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const today = new Date().toISOString().split('T')[0];
    const activeOnly = (req.query.active || 'true').toLowerCase() === 'true';
    const donViIds = await getDonViChildIds(dvid);
    if (!donViIds.length) return res.json([]);
    const ph = donViIds.map(() => '?').join(',');
    let records;
    if (activeOnly) {
      records = await db.all(`SELECT cc.* FROM cat_com cc JOIN chien_si cs ON cc.chien_si_id = cs.id
        WHERE cs.don_vi_id IN (${ph}) AND cc.ngay_bat_dau <= ? AND cc.ngay_ket_thuc >= ? ORDER BY cc.ngay_tao DESC`, ...donViIds, today, today);
    } else {
      records = await db.all(`SELECT cc.* FROM cat_com cc JOIN chien_si cs ON cc.chien_si_id = cs.id
        WHERE cs.don_vi_id IN (${ph}) ORDER BY cc.ngay_tao DESC`, ...donViIds);
    }
    res.json(await Promise.all(records.map(catComToDict)));
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.post('/api/meal-cuts', loginRequired, async (req, res) => {
  try {
    const { chien_si_id, ngay_bat_dau, ngay_ket_thuc, ly_do = '', loai = 'cat_com', ghi_chu, nguoi_bao } = req.body;
    const r = await db.run('INSERT INTO cat_com (chien_si_id, ngay_bat_dau, ngay_ket_thuc, ly_do, loai, ghi_chu, nguoi_bao) VALUES (?, ?, ?, ?, ?, ?, ?)',
      chien_si_id, ngay_bat_dau, ngay_ket_thuc, ly_do, loai, ghi_chu || null, nguoi_bao || null);
    const cs = await db.get('SELECT * FROM chien_si WHERE id = ?', chien_si_id);
    if (cs) {
      const newStatus = ly_do === 'phep' ? 'phep' : (ly_do === 'gac' ? 'gac' : 'cong_tac');
      await db.run('UPDATE chien_si SET trang_thai = ? WHERE id = ?', newStatus, chien_si_id);
    }
    const cc = await db.get('SELECT * FROM cat_com WHERE id = ?', r.lastInsertRowid);
    res.status(201).json(await catComToDict(cc));
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.delete('/api/meal-cuts/:cutId', loginRequired, async (req, res) => {
  try {
    const cc = await db.get('SELECT * FROM cat_com WHERE id = ?', req.params.cutId);
    if (!cc) return res.status(404).json({ error: 'Not found' });
    await db.run("UPDATE chien_si SET trang_thai = 'tai_vi' WHERE id = ?", cc.chien_si_id);
    await db.run('DELETE FROM cat_com WHERE id = ?', cc.id);
    res.json({ message: 'Đã hủy cắt cơm' });
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.get('/api/personnel/birthdays', loginRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const month = parseInt(req.query.month) || (new Date().getMonth() + 1);
    const donViIds = await getDonViChildIds(dvid);
    if (!donViIds.length) return res.json([]);
    const ph = donViIds.map(() => '?').join(',');
    const IS_PG = !!process.env.DATABASE_URL;
    let list;
    if (IS_PG) {
      list = await db.all(`SELECT * FROM chien_si WHERE EXTRACT(MONTH FROM ngay_sinh::date)::integer = ? AND don_vi_id IN (${ph})`, month, ...donViIds);
    } else {
      list = await db.all(`SELECT * FROM chien_si WHERE CAST(strftime('%m', ngay_sinh) AS INTEGER) = ? AND don_vi_id IN (${ph})`, month, ...donViIds);
    }
    res.json(await Promise.all(list.map(chienSiToDict)));
  } catch (e) { res.status(500).json({ error: e.message }); }
});

module.exports = router;
